### ISSUE TITLE
Add method for T2* estimation/correction from complex data

### ISSUE THREAD

**Issue by tsalo** [ME-ICA/tedana#531]
### Summary
Per the resources that @agahkarakuzu kindly shared with us in https://github.com/ME-ICA/tedana/issues/464#issuecomment-583893763, T2* estimates should be corrected for background field gradients using the phase data.

### Additional Detail
There is a MATLAB implementation [here](https://github.com/neuropoly/t2star-mapping), as well as in [qMRLab](https://github.com/qMRLab/qMRLab), that we could translate to Python for tedana.

### Next Steps
1. Translate T2* correction method to Python.
2. Test against original MATLAB version with ME-GRE data, given that high-resolution GRE data with many echoes is what the method was developed on.
3. Apply to ME-EPI data with much fewer echoes (e.g., 4) to determine if the method works well with ME-fMRI.
4. Support in t2smap and tedana workflows.
5. Add documentation.
6. Add tests.

**Comment by tsalo** [comment]
@dowdlelt I only have access to complex, four-echo data. Is there any chance that you reconstructed phase data for any of the eight-echo datasets you've run and that you could share those data for testing this feature?

Alternatively, if anyone has any ME-GRE data we could use, that would also work.

**Comment by dowdlelt** [comment]
I did not, unfortunately. Just had magnitude. I'll keep this in mind if I get in the scanner to do some testing, but its not super likely in the short term.

**Comment by agahkarakuzu** [comment]
@tsalo I’ll get an example dataset soon, I think I can share with you. As a side note, processing time is going to be quite long. BTW is it not possible to obtain phase images from the complex data you have in hand, using e.g. ‘fslcomplex’ module?

**Comment by tsalo** [comment]
@agahkarakuzu That would be amazing!

I do have phase data, but only four echoes. In order to test out the method as I manage to translate it to Python, I'd prefer to use some data with a relatively large number of echoes.

**Comment by agahkarakuzu** [comment]
@tsalo 4 echoes are not that bad btw, the MEGRE data I’ll have is also 4 echoes. Have you visually inspected the decay curve of your magnitude images across 4 echoes?

**Comment by tsalo** [comment]
Ah, but my data are ME-EPI, not GRE, so it's relatively low-resolution. Four-echo MEGRE data would definitely be better for testing the code.

I haven't looked at the signal decay for my data yet, but I recall our multi-echo fMRI denoising did a pretty good job on it.

**Comment by agahkarakuzu** [comment]
@tsalo asking out of curiosity, do you have an example acquisition protocol for ME-EPI (a PDF)? I did some reading after hearing about ME-EPI from @emdupre few months ago, because from the pulse sequence perspective there is not a single-echo EPI and I wondered what was meant by "ME" in terms of sampling the data :) 

I found that paper I skimmed (by Kundu et al.), is your T2* sampling similar to the following? 
![image](https://user-images.githubusercontent.com/9632322/75207878-fa308600-5747-11ea-80cf-80fb63561ddd.png)

If so, 4 echoes (if split enough) would sample decent decay for parameter estimation. Ofc, not the ideal dataset for testing the code after matlab --> python translation. I can give you the fitted map as well for the comparison, it should save you some time. 

I am curious to see how mono-exponential T2* fitting with ∆Bz correction is going to differ from the one without correction for ME-EPI data.

**Comment by tsalo** [comment]
I actually just recently posted a PDF for the four-echo protocol I'm using for my dissertation's sample [here](https://osf.io/y46cu/). It uses the CMRR MB sequence (specifically R016) for Siemens.

>I found the paper I skimmed (by Kundu et al.), is your T2* sampling similar to the following?

Yes. `tedana` is actually a heavily updated version of Kundu's multi-echo fMRI T2* estimation and echo-dependence denoising package. We have both log-linear and nonlinear mono-exponential T2* fitting implemented, but no ∆Bz correction just yet. I'm pretty excited to see how it comes out.

**Comment by agahkarakuzu** [comment]
@tsalo thanks, really cool dataset!

**Comment by agahkarakuzu** [comment]
I uploaded the dataset [here](https://osf.io/y4qs9/?view_only=94d0ea023b9747c9a1ed0e3d30b5c5bb). I hope you are all doing good, stay safe!

**Comment by tsalo** [comment]
I come back to this every once in a while, and I'm thinking of carving some time out to try translating the MATLAB implementation in https://github.com/qMRLab/qMRLab/pull/445 into Python for tedana.

**Comment by tsalo** [comment]
I've got a working Python version written by Codex. The method modifies the magnitude data using the phase data, and the phase data needs to be in native space, so it may make more sense to have this as a separate CLI that can be run in the following pipeline:

1. Run preprocessing, retaining transforms.
2. Unwrap raw phase data.
3. Apply correction to raw data.
4. Apply transforms to corrected magnitude data.
5. Estimate T2* and S0 from the corrected data using `t2smap`.
6. Pass T2*/S0 from corrected data into `tedana`, but use echo-wise preprocessed (uncorrected) data as the inputs, as the correction is probably not appropriate for general processing.

**Comment by agahkarakuzu** [comment]
This is so fascinating @tsalo! I was experimenting with matlab --> rust conversion using Claude, impressively reliable.

**Comment by tsalo** [comment]
Indeed! I used it to translate my Python translation of NORDIC (plus the original MATLAB code) to Rust and it got *so much faster*!

**Comment by tsalo** [comment]
Here is the Python translation BTW: https://github.com/ME-ICA/dahnke

Here's an example of how the correction tends to affect T2* estimation. T2* is in ms. I estimated T2* for one of the runs in my PAFIN dataset with and without the correction.

<img width="1800" height="600" alt="Image" src="https://github.com/user-attachments/assets/5dfe1c83-2381-42e1-b37c-3d13120dc53c" />

EDIT: I didn't realize this before (it's not part of the t2star-mapping code), but we can also directly use the complex-valued data to estimate T2*. See Eq. 1 in https://pmc.ncbi.nlm.nih.gov/articles/PMC3893831/.
